usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,249 @@
+
+	if regularFileExists(desktopFile) {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (not all entries in $XDG_DATA_DIRS) and we know that snapd
+	// does not create subdirectories in that location.
+	for i := 1; i != len(splitFileId); i++ {
+		desktopFile, err := findDesktopFile(filepath.Join(baseDir, strings.Join(splitFileId[:i], "-")), splitFileId[i:])
+		if err == nil {
+			return desktopFile, err
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileID string) (string, error) {
+	// OpenDesktopEntry() currently only supports launching snap applications from
+	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
+	// written by snapd and considered safe for userd to process.
+	// Since we only access /var/lib/snapd/desktop/applications, ignore
+	// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	baseDir := dirs.SnapDesktopFilesDir
+
+	desktopFile, err := findDesktopFile(baseDir, strings.Split(desktopFileID, "-"))
+
+	if err == nil {
+		return desktopFile, nil
+	}
+
+	return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
+}
+
+// verifyDesktopFileLocation checks the desktop file location:
+// we only consider desktop files in dirs.SnapDesktopFilesDir
+func verifyDesktopFileLocation(desktopFile string) error {
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir+"/") {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("internal error: only launching snap applications from /var/lib/snapd/desktop/applications is supported")
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (exec string, icon string, err error) {
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return exec, icon, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	inDesktopSection := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
```

Comment:
this implementation seems to suffer from either a bug or a feature regarding multiple `"[Desktop Entry"]` lines, i.e. if I give it something like this:

```
[Desktop Entry]
Exec=good
[Desktop Entry]
Exec=bad
```
then we seem to find exec as "bad", and it's unclear if this is okay or not. Arguably it is because these desktop files are _supposed_ to be written and sanitized etc by snapd, but bugs are a thing so I guess I would feel better if this just straight failed if we see multiple Deskop Entry lines (and same for multiple Exec and Icon lines).

Reply:
in fact reading through our implementation of sanitizeDesktopFile, we don't actually do anything about duplicated `[Desktop Entry]` sections so I think we definitely need to fail in that case here (in addition to fixing the bug where we allow multiple `Deskop Entry` sections in through

Reply:
For what it is worth, the desktop entry spec specifies that ["Multiple groups may not have the same name"](https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s03.html#group-header), and the GKeyFile code in GLib errors out if it sees duplicates.  So this would be worth doing.

Something as simple as adding a `seenDesktopSection` boolean variable should do, so we can detect the duplicate section.